### PR TITLE
[NVPTX] Fix NVPTXLowerUnreachable::isLoweredToTrap logic

### PR DESCRIPTION
**Pull request by duk-37**
Previously, this pass would not generate traps if `NoTrapAfterNoreturn` was set and would generate traps even if the instruction directly before the `UnreachableInst` was `llvm.trap()`.

Fix both of these problems and add some tests.

### CODE REVIEW

**Review by chsigg — APPROVED** [review]
Thanks!

**Comment by duk-37** [comment]
FYI I don't have commit access anymore so you'll have to merge it in. :)

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp [hunk 1]
```diff
@@ -110,17 +110,25 @@ StringRef NVPTXLowerUnreachable::getPassName() const {
...
 // This is a copy of the logic in SelectionDAGBuilder::visitUnreachable().
 // =============================================================================
 bool NVPTXLowerUnreachable::isLoweredToTrap(const UnreachableInst &I) const {
-  if (!TrapUnreachable)
-    return false;
-  if (!NoTrapAfterNoreturn)
-    return true;
   const CallInst *Call = dyn_cast_or_null<CallInst>(I.getPrevNode());
-  return Call && Call->doesNotReturn();
+  if (Call) {
```
Comment (justinfargnoli) [inline llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:119]: Nit: move the definition of `Call` into the `if` condition or [prefer early returns](https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code)

**Comment by justinfargnoli** [comment]
> this pass ... would generate traps even if the instruction directly before the UnreachableInst was llvm.trap()

Why is this a problem?

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp [hunk 2]
```diff
@@ -110,17 +110,25 @@ StringRef NVPTXLowerUnreachable::getPassName() const {
...
+    // We've already emitted a non-continuable trap.
+    if (Call->isNonContinuableTrap())
+      return true;
```
Comment (justinfargnoli) [inline llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:122]: > this pass ... would generate traps even if the instruction directly before the UnreachableInst was llvm.trap()

Why is this a problem?

Comment (duk-37) [reply llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:122]: It's not in the sense that there's no miscompile involved, but correctly matching the logic in the instruction selectors happens to change this case too.

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp [hunk 3]
```diff
@@ -110,17 +110,25 @@ StringRef NVPTXLowerUnreachable::getPassName() const {
...
+
+    // No traps are emitted for calls that do not return
+    // when this option is enabled.
+    if (NoTrapAfterNoreturn && Call->doesNotReturn())
```
Comment (justinfargnoli) [inline llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:126]: Nit: rename `NoTrapAfterNoreturn` to `NoTrapAfterNoReturn`

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp [hunk 4]
```diff
@@ -110,17 +110,25 @@ StringRef NVPTXLowerUnreachable::getPassName() const {
...
-// Returns whether a `trap` intrinsic should be emitted before I.
+// Returns whether a `trap` intrinsic would be emitted before I.
 //
 // This is a copy of the logic in SelectionDAGBuilder::visitUnreachable().
 // =============================================================================
 bool NVPTXLowerUnreachable::isLoweredToTrap(const UnreachableInst &I) const {
-  if (!TrapUnreachable)
-    return false;
-  if (!NoTrapAfterNoreturn)
-    return true;
```
Comment (justinfargnoli) [inline llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:121]: > this pass would not generate traps if NoTrapAfterNoreturn was set

It seems like a trap would be generated if `NoTrapAfterNoreturn` is set and `Call->doesNotReturn()` is true. Am I missing something?

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp (hunk 2 above)
Comment (justinfargnoli) [reply llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:122]: Can you provide a link to where this logic malfunctions in the instruction selector?

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp (hunk 4 above)
Comment (duk-37) [reply llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:121]: AIUI `isLoweredToTrap` is trying to detect whether a trap would _already_ be generated by isel for any given unreachable instruction. It does this by replicating the logic for emitting traps in `SelectionDAGBuilder::visitUnreachable`. That's why this function returning true if  `NoTrapAfterNoreturn && Call->doesNotReturn()` is wrong

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp (hunk 2 above)
Comment (duk-37) [reply llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:122]: The instruction selector is correct -- logic is at `SelectionDAGBuilder::visitUnreachable`

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp (hunk 3 above)
Comment (duk-37) [reply llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:125]: I don't think renaming fields would be appropriate for this commit, though there are definitely some parts of the pass that could use some reformatting. Also, `NoTrapAfterNoreturn` matches casing of the `TargetOptions` field so the current name feels appropriate unless we change that too, IMHO

**Review by justinfargnoli — APPROVED** [review]
LGTM :)

**Pull request merged**